src/Utilities/Compiler/WellKnownTypeProvider.cs

```diff
@@ -46,7 +48,47 @@ public bool TryGetOrCreateTypeByMetadataName(string fullTypeName, out INamedType
         {
             namedTypeSymbol = _fullNameToTypeMap.GetOrAdd(
                 fullTypeName,
-                (string s) => Compilation.GetTypeByMetadataName(s));    // Caching null results in our cache is intended.
+                fullyQualifiedMetadataName =>
+                {
+                    // Caching null results in our cache is intended.
+                    var type = Compilation.GetTypeByMetadataName(fullyQualifiedMetadataName)
+                        ?? Compilation.Assembly.GetTypeByMetadataName(fullyQualifiedMetadataName);
+                    if (type is null)
+                    {
+                        foreach (var module in Compilation.Assembly.Modules)
+                        {
+                            foreach (var referencedAssembly in module.ReferencedAssemblySymbols)
```

Comment:
I am not sure whether there is a real impact or anything but I thought that it was better to use `for` loop instead of `foreach` loop for `ImmutableArray`?

Reply:
ImmutableArray has a struct enumerator, so it should be identical.